src/vmm/src/devices/virtio/iov_deque.rs

```diff
@@ -0,0 +1,412 @@
+    ///
+    /// This will allocate exactly two pages of virtual memory. In order to implement the
+    /// optimization that allows us to always have elements in contiguous memory we need
+    /// allocations at the granularity of `PAGE_SIZE`. Now, our queues are at maximum 256
+    /// descriptors long and `struct iovec` looks like this:
+    ///
+    /// ```Rust
+    /// pub struct iovec {
+    ///    pub iov_base: *mut ::c_void,
+    ///    pub iov_len: ::size_t,
+    /// }
+    /// ```
+    ///
+    /// so, it's 16 bytes long. As a result, we need a single page for holding the actual data of
+    /// our buffer.
+    fn allocate_ring_buffer_memory() -> Result<*mut c_void, IovDequeError> {
+        // The fact that we allocate two pages is due to the size of `struct iovec` times our queue
+        // size equals the page size. Add here a debug assertion to reflect that and ensure that we
+        // will adapt our logic if the assumption changes in the future.
+        debug_assert_eq!(
+            std::mem::size_of::<iovec>() * usize::from(FIRECRACKER_MAX_QUEUE_SIZE),
+            PAGE_SIZE
+        );
+
+        Self::mmap(
+            std::ptr::null_mut(),
+            PAGE_SIZE * 2,
+            libc::PROT_NONE,
+            libc::MAP_PRIVATE | libc::MAP_ANONYMOUS,
+            -1,
+            0,
+        )
+    }
+
+    /// Create a new [`IovDeque`] that can hold memory described by a single VirtIO queue.
+    pub(crate) fn new() -> Result<Self, IovDequeError> {
+        let memfd = Self::create_memfd()?;
+        let raw_memfd = memfd.as_file().as_raw_fd();
+        let buffer = Self::allocate_ring_buffer_memory()?;
+
+        // Map the first page of virtual memory to the physical page described by the memfd object
+        let _ = Self::mmap(
+            buffer,
+            PAGE_SIZE,
+            libc::PROT_READ | libc::PROT_WRITE,
+            libc::MAP_SHARED | libc::MAP_FIXED,
+            raw_memfd,
+            0,
+        )?;
+
+        // Map the second page of virtual memory to the physical page described by the memfd object
+        // SAFETY: safe because `Self::allocate_ring_buffer_memory` allocates exactly two pages for
+        // us
+        let next_page = unsafe { buffer.add(PAGE_SIZE) };
+        let _ = Self::mmap(
+            next_page,
+            PAGE_SIZE,
+            libc::PROT_READ | libc::PROT_WRITE,
+            libc::MAP_SHARED | libc::MAP_FIXED,
+            raw_memfd,
+            0,
+        )?;
+
+        // SAFETY:
+        // * `buffer` is valid both for reads and writes (allocated with `libc::PROT_READ |
+        //   libc::PROT_WRITE`. `
+        // * `buffer` is aligned at `PAGE_SIZE`
+        // * `buffer` points to memory allocated with a single system call to `libc::mmap`
+        let iov = unsafe {
+            std::slice::from_raw_parts_mut(
+                buffer.cast(),
+                2 * PAGE_SIZE / std::mem::size_of::<libc::iovec>(),
+            )
+        };
```

Comment:
Mhhh, I don't think we can do this. The two pages contain the same physical memory, so we're creating aliasing mutable references here. But if we just store everything as `*mut libc::iovec`, I think it'll be fine :)

Reply:
And if we use pointers, we also get rid of the lifetime parameter (which is kinda wrong to have anyway, because the `iov` field is self referential, and that's not something rust's lifetimes can express)

Reply:
That was a good catch. I updated the code to only deal with pointers.